### DISCUSSION TITLE
Undefined `limit` and `offset` are not compatible with `FindOptions`

### DISCUSSION THREAD

**Discussion by peschee** [mikro-orm/mikro-orm#5122] [General]
### Describe the bug

The documentation has the following example:

```typescript
app.get('/article', async request => {
  const { limit, offset } = request.query as { limit?: number; offset?: number };
  const [items, total] = await orm.em.findAndCount(Article, {}, {
    limit, offset,
  });

  return { items, total };
});
```

meaning `limit` and `offset` can be `undefined`. However, the `FindOptions` interface used here does not allow `limit` and `offset` to be undefined: 

https://github.com/mikro-orm/mikro-orm/blob/d0d5de8cc0b0f290a75dbd5962953b8e4065d02e/packages/core/src/drivers/IDatabaseDriver.ts#L116

so you end up with the following TS error:

```
TS2379: Argument of type
{     limit: number | undefined;     offset: number | undefined; }
is not assignable to parameter of type FindOptions<Event, never, "*", never> with 'exactOptionalPropertyTypes: true'. Consider adding undefined to the types of the target's properties.
  Types of property limit are incompatible.
    Type number | undefined is not assignable to type number
      Type undefined is not assignable to type number
```

### Reproduction

No

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.0.2

### Node.js version

20.10.0

### Operating system

MacOS 14.2.1 (23C71)

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by peschee** [comment]
Oh no, reading this I see in the link I posted that `limit` CAN in fact be undefined: `limit?: number;`

Hmm…

**Comment by peschee** [comment]
Ok, I have a small repro here: https://github.com/peschee/mikro-orm-6-postgres-migrations-test/tree/bugfix/limit-offset

Running `npm run build` renders the error described:

```shell
src/index.ts:10:3 - error TS2379: Argument of type '{ limit: undefined; offset: undefined; }' is not assignable to parameter of type 'FindOptions<Event, never, "*", never>' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the types of the target's properties.
  Types of property 'limit' are incompatible.
    Type 'undefined' is not assignable to type 'number'.

10   {
     ~
11     limit,
   ~~~~~~~~~~
12     offset,
   ~~~~~~~~~~~
13   }
   ~~~
```

**Reply by B4nan** [reply]
> exactOptionalPropertyTypes

Hmm first time hearing about this, not sure if I want to adopt this, feels like this would require many more changes and the final code would be rather ugly. This would affect everything, not just the limit/offset fields, including entity definition.

But it feels quite weird that this option would require existing projects to adopt it first for you to use it.

(I don't mind adding the `| undefined` for the few props in `FindOptions`, we can surely do that, PR welcome, but it feels like you will soon hit something similar elsewhere, and I can imagine other packages will be similarly problematic.)

**Reply by peschee** [reply]
Ah ok, I see. The `"exactOptionalPropertyTypes": true` comes from [`@tsconfig/strictest/tsconfig`](https://www.npmjs.com/package/@tsconfig/strictest).

Setting this option to `false` solves the issue. Good enough for me, I can live without it, although the [docs](https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes) state:

> `exactOptionalPropertyTypes` makes TypeScript truly enforce the definition provided as an optional property:

which also sounds "good" :)

**Reply by B4nan** [reply]
Yeah, I will take a look at how enabling this flag behaves in some of the more complex example repos, if it's just about the FindOptions we can surely make it work, I do want to be compatible with this generally (but I don't personally want to use that in the repo itself).
